### ISSUE TITLE
Secure uploads

### ISSUE THREAD

**Issue by homm** [uploadcare/uploadcare-widget#308]
"Secure uploads" is the method to protect uploads from using without a secret key. All uploads should be signed if appropriate option is turned on for the project.
- If a signature is provided, Widget should send signed credentials with each upload request:
  - [from_url](https://github.com/uploadcare/uploadcare-widget/blob/master/app/assets/javascripts/uploadcare/files/url.coffee#L51)
  - [base upload](https://github.com/uploadcare/uploadcare-widget/blob/master/app/assets/javascripts/uploadcare/files/object.coffee#L97)
  - [multipart start](https://github.com/uploadcare/uploadcare-widget/blob/master/app/assets/javascripts/uploadcare/files/object.coffee#L157)
  - [iframe upload](https://github.com/uploadcare/uploadcare-widget/blob/master/app/assets/javascripts/uploadcare/files/input.coffee#L18)
- There should be some callback on the host page which will be called if signed credentials are needed. For example for first uploading or if some uploading is failed due to credentials checking failure.
- The widget must "understand" if upload is failed due to credentials checking failure, call the callback again and retry uploading.
- Credentials should be shared between the widget instances and callback shouldn't be requested more times than enough. For example, when 5 uploads have failed due to credentials checking failure, the callback should be called only once.
- It is should be possible to use different credentials with different public keys.

Usage example:

``` html
<script>
    UPLOADCARE_PUBLIC_KEY = 'ba4a62161fc08614d93d';
    uploadcare.registerSecureUploadsCallback(UPLOADCARE_PUBLIC_KEY, function() {
      var deferred = $.Deferred();
      $.ajax({
        'url': '/uploadcare_secure_uploads_credentials/',
        'type': 'json',
      }).done(function(result) {
        deferred.resolve(result.expire, result.sign);
      });
      return deferred.promise();
    });
</script>
```

Alternative simple way without callback for long-living signatures:

``` html
<script>
    UPLOADCARE_PUBLIC_KEY = 'ba4a62161fc08614d93d';
    UPLOADCARE_SECURE_EXPIRES = 1487078477;
    UPLOADCARE_SECURE_SIGNATURE = 'afe1079beec9a281f622c7329b53bf7c';
</script>
```

**Comment by fabien** [comment]
+1

**Comment by omin** [comment]
Thanks. I couldn't find this in the documentation.

**Comment by omin** [comment]
Actually, has this been fixed? Adding my `UPLOADCARE_SECURE_EXPIRES` and `UPLOADCARE_SECURE_SIGNATURE` did not work for me.

**Comment by dmitry-mukhin** [comment]
@omin, how are you generating those signatures?
Please read more on this here: https://uploadcare.com/documentation/upload/#secure-uploads

If it does not help, please contact us via email.

**Comment by omin** [comment]
@dmitry-mukhin exactly as the document. Double checked the md5 hashing outputs as well.

**Comment by dmitry-mukhin** [comment]
let's move to email. this is an implementation issue, not a widget issue.

**Comment by omin** [comment]
@dmitry-mukhin you were right.  I found a problem on the frontend.
